File: apps/api/internal/application/dtos/capability.go
package dtos

// Capability request/response DTOs for the admin API (PRD §59-64).
// Secrets are never exposed — only the credential_reference string (PRD §61, §91).

// CreateCapabilityRequest is the payload to register a capability.
type CreateCapabilityRequest struct {
	Name                string `json:"name"`
	Description         string `json:"description"`
	ProviderType        string `json:"providerType"`
	ProviderID          string `json:"providerId"`
	InputSchema         any    `json:"inputSchema"`
	OutputSchema        any    `json:"outputSchema"`
	Version             int    `json:"version"`
	CredentialReference string `json:"credentialReference"`
}

// UpdateCapabilityRequest is the payload to update a capability's mutable fields.
type UpdateCapabilityRequest struct {
	Description         string `json:"description"`
	ProviderType        string `json:"providerType"`
	ProviderID          string `json:"providerId"`
	InputSchema         any    `json:"inputSchema"`
	OutputSchema        any    `json:"outputSchema"`
	Status              string `json:"status"`
	Version             int    `json:"version"`
	CredentialReference string `json:"credentialReference"`
}

// CapabilityDTO is the serializable capability returned to callers. It exposes
// only credential_reference, never the resolved secret.
type CapabilityDTO struct {
	ID                  string  `json:"id"`
	TenantID            string  `json:"tenantId"`
	Name                string  `json:"name"`
	Description         *string `json:"description"`
	ProviderType        string  `json:"providerType"`
	ProviderID          *string `json:"providerId"`
	InputSchema         any     `json:"inputSchema"`
	OutputSchema        any     `json:"outputSchema"`
	Status              string  `json:"status"`
	Version             int     `json:"version"`
	CredentialReference *string `json:"credentialReference"`
	CreatedAt           string  `json:"createdAt"`
	UpdatedAt           string  `json:"updatedAt"`
}

// CapabilityListDTO wraps a tenant-scoped capability list.
type CapabilityListDTO struct {
	Data []CapabilityDTO `json:"data"`
}

// CreateBindingRequest binds a capability to a tenant/workflow/state scope.
type CreateBindingRequest struct {
	ScopeType  string `json:"scopeType"`
	ScopeID    string `json:"scopeId"`
	Permission string `json:"permission"`
}

// CapabilityBindingDTO is the serializable capability binding.
type CapabilityBindingDTO struct {
	ID           string `json:"id"`
	TenantID     string `json:"tenantId"`
	CapabilityID string `json:"capabilityId"`
	ScopeType    string `json:"scopeType"`
	ScopeID      string `json:"scopeId"`
	Permission   string `json:"permission"`
	CreatedAt    string `json:"createdAt"`
	UpdatedAt    string `json:"updatedAt"`
}

// TestInvocationRequest is the payload to test-invoke a capability in
// sandbox/mock mode (PRD §2064).
type TestInvocationRequest struct {
	Payload map[string]any `json:"payload"`
	ScopeID string         `json:"scopeId"` // optional workflow/state scope for binding resolution
}

// TestInvocationResultDTO is the normalized result flagged fromMock.
type TestInvocationResultDTO struct {
	Data       map[string]any `json:"data"`
	FromMock   bool           `json:"fromMock"`
	DurationMS int64          `json:"durationMs"`
	Event      *string        `json:"event"`
}

// CapabilityErrorDTO is a classified capability failure (PRD §87).
type CapabilityErrorDTO struct {
	Kind    string `json:"kind"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

File: packages/schemas/capability.ts
import { z } from "zod"

// Provider type of a capability (PRD §59).
export const providerTypes = ["MCP", "INTERNAL", "HTTP", "FUTURE"] as const

export const providerTypeLabels = [
  { label: "MCP", value: "MCP" },
  { label: "Internal", value: "INTERNAL" },
  { label: "HTTP", value: "HTTP" },
  { label: "Future", value: "FUTURE" },
]

export const getProviderTypeLabel = (value: (typeof providerTypes)[number]) => {
  return providerTypeLabels.find((l) => l.value === value)?.label ?? value
}

// Lifecycle of a capability in the registry.
export const capabilityStatuses = ["ACTIVE", "INACTIVE", "DISABLED"] as const

export const capabilityStatusLabels = [
  { label: "Active", value: "ACTIVE" },
  { label: "Inactive", value: "INACTIVE" },
  { label: "Disabled", value: "DISABLED" },
]

export const getCapabilityStatusLabel = (
  value: (typeof capabilityStatuses)[number],
) => {
  return capabilityStatusLabels.find((l) => l.value === value)?.label ?? value
}

// Scope type of a capability binding (PRD §60).
export const bindingScopeTypes = ["TENANT", "WORKFLOW", "STATE"] as const

export const bindingScopeTypeLabels = [
  { label: "Tenant", value: "TENANT" },
  { label: "Workflow", value: "WORKFLOW" },
  { label: "State", value: "STATE" },
]

export const getBindingScopeTypeLabel = (
  value: (typeof bindingScopeTypes)[number],
) => {
  return bindingScopeTypeLabels.find((l) => l.value === value)?.label ?? value
}

// Permission of a capability binding.
export const bindingPermissions = ["ALLOW", "DENY"] as const

export const bindingPermissionLabels = [
  { label: "Allow", value: "ALLOW" },
  { label: "Deny", value: "DENY" },
]

export const getBindingPermissionLabel = (
  value: (typeof bindingPermissions)[number],
) => {
  return bindingPermissionLabels.find((l) => l.value === value)?.label ?? value
}

// Free-form JSON object (input/output schema, payload). Optional values default
// to an empty object.
const jsonSchemaField = z.record(z.string(), z.unknown()).optional()

// Register a capability (PRD §59).
export const createCapabilitySchema = z.object({
  name: z.string().min(1, "Name is required"),
  description: z.string().optional(),
  providerType: z.enum(providerTypes, {
    errorMap: () => ({ message: "Provider type is required" }),
  }),
  providerId: z.string().optional(),
  providerTool: z.string().optional(),
  inputSchema: jsonSchemaField,
  outputSchema: jsonSchemaField,
  version: z.number().int().positive().optional(),
  credentialReference: z.string().optional(),
})

export type CreateCapabilitySchemaProps = z.infer<typeof createCapabilitySchema>

// Update a capability's mutable fields (partial/PATCH).
export const updateCapabilitySchema = z.object({
  description: z.string().optional(),
  providerType: z.enum(providerTypes).optional(),
  providerId: z.string().optional(),
  providerTool: z.string().optional(),
  inputSchema: jsonSchemaField,
  outputSchema: jsonSchemaField,
  status: z.enum(capabilityStatuses).optional(),
  version: z.number().int().positive().optional(),
  credentialReference: z.string().optional(),
})

export type UpdateCapabilitySchemaProps = z.infer<typeof updateCapabilitySchema>

// Bind a capability to a tenant/workflow/state scope (PRD §60).
export const bindingSchema = z.object({
  scopeType: z.enum(bindingScopeTypes, {
    errorMap: () => ({ message: "Scope type is required" }),
  }),
  scopeId: z.string().min(1, "Scope id is required"),
  permission: z.enum(bindingPermissions).optional(),
})

export type BindingSchemaProps = z.infer<typeof bindingSchema>

// Test-invoke a capability in sandbox/mock mode (PRD §2064).
export const testInvocationSchema = z.object({
  payload: z.record(z.string(), z.unknown()).default({}),
  scopeId: z.string().optional(),
})

export type TestInvocationSchemaProps = z.infer<typeof testInvocationSchema>
